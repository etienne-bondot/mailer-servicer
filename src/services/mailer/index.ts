import emailTemplate from 'email-templates';
import nodemailer, {Transporter, Transport, TransportOptions} from 'nodemailer';

import {Options} from '~/types';

type Provider = 'gmail';

export class Mailer {
  // Services
  transporter: Transporter;
  mailer: emailTemplate;

  // Options
  options: Options;

  constructor({
    provider,
    options,
    transporterOptions,
  }: {
    provider?: Provider;
    options?: Options;
    transporterOptions?: TransportOptions;
  }) {
    this.options = {
      isTesting: false,
      verbose: true,
      ...options,
    };

    function getTransporter(provider?: Provider): TransportOptions | Transport {
      switch (provider) {
        case 'gmail':
        default:
          return {
            // @ts-expect-error 'host' does not exist in type 'TransportOptions | Transport'
            host: 'smtp.gmail.com',
            port: 465,
            secure: true,
            service: 'gmail',
            auth: {
              type: 'OAuth2',
              user: process.env.GMAIL_USERNAME,
              clientId: process.env.OAUTH_CLIENT_ID,
              clientSecret: process.env.OAUTH_CLIENT_SECRET,
              refreshToken: process.env.OAUTH_REFRESH_TOKEN,
              accessToken: process.env.OAUTH_ACCESS_TOKEN,
              expires: 3599,
            },
          };
      }
    }

    this.transporter = nodemailer.createTransport(transporterOptions || getTransporter(provider));

    this.mailer = new emailTemplate({
      send: true,
      preview: false,
      views: {root: './emails'},
      transport: this.options.isTesting ? {jsonTransport: true} : this.transporter,
    });
  }

  async send({
    subject,
    to,
    template,
    data,
  }: {
    subject: string;
    to: string;
    template: string;
    data?: {[key: string]: string | number};
  }): Promise<{message?: string; data?: {[key: string]: unknown}; success: boolean}> {
    try {
      const response = await this.mailer.send({
        template,
        message: {
          subject,
          to,
        },
        locals: data || {},
      });

      if (this.options.verbose) {
        console.log({data: response});
      }
      return {data: response, success: true};
    } catch (err) {
      if (this.options.verbose) {
        console.error(err);
      }
      return {message: err, success: false};
    }
  }
}

let mailer;
const provider = process.env.PROVIDER;

try {
  if (!provider) {
    throw new Error(
      'No provider set - add PROVIDER=<your_provider> to the .env file (actual providers supported: "gmail")'
    );
  } else {
    console.log(`Setting new provider: ${provider}`);
    mailer = new Mailer({provider: provider as Provider});
  }
} catch (error) {
  console.error(error);
}

export {mailer};

export default Mailer;
