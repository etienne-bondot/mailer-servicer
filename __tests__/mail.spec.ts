import {Mailer} from '../src/services/mailer';

const verbose = false;

const testResponse = (response, data) => {
  expect(response).toHaveProperty('success');
  expect(response.success).toBeTruthy();
  expect(response).toHaveProperty('data');
  expect(response.data).toHaveProperty('message');
  const message = JSON.parse(response.data.message);
  expect(message).toHaveProperty('subject');
  expect(message).toHaveProperty('html');
  expect(message).toHaveProperty('text');
  expect(message.subject).toEqual(data.subject);
  expect(message.html).toMatchSnapshot();
  expect(message.text).toMatchSnapshot();
};

const templates = {
  'testing-template': {
    subject: 'Testing',
    data: {
      link: '/test/link',
      name: 'Username',
      token: 'test_token',
    },
  },
};

const templateNames = Object.keys(templates);

describe('Templates', () => {
  const credentials = {
    username: 'fake@gmail.com',
    clientId: 'fakeClientId',
    clientSecret: 'fakeClientSecret',
    refreshToken: 'fakeRefreshToken',
    accessToken: 'fakeAccessToken',
  };

  for (const templateName of templateNames) {
    const template = templates[templateName];

    it(templateName, async () => {
      // given
      const mailer = new Mailer({
        options: {
          isTesting: true,
          verbose,
        },
        transporterOptions: {
          // @ts-expect-error 'host' does not exist in type 'TransportOptions | Transport'
          host: 'smtp.gmail.com',
          port: 465,
          secure: true,
          service: 'gmail',
          auth: {
            ...credentials,
            type: 'OAuth2',
            expires: 3599,
          },
        },
      });

      // when
      const res = await mailer.send({
        subject: template.subject,
        to: 'email+test@fakeaddress.com',
        template: templateName,
        data: template.data,
      });

      // then
      testResponse(res, template);
    });
  }
});
