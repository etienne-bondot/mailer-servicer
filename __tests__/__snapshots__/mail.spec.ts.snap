// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Templates testing-template 1`] = `
"
<p>Hi Username,</p>
<p>This is a testing template with a fake link</p><a href=\\"/test/link?token=test_token\\">Fake Link</a>
<p>Cheers!</p>"
`;

exports[`Templates testing-template 2`] = `
"Hi Username,

This is a testing template with a fake link

Fake Link [/test/link?token=test_token]

Cheers!"
`;
